[ios/tvos] Refactor build/cmake scripts, implement codesigning arguments for xcode project generation

fuzzard opened the pull request:
## Description
Refactor and cleanup several cmake scripts and build scripts for ios/tvos platforms
Implement codesigning arguments from command line to populate xcode project attributes


## Motivation and Context
Looking to start upstreaming some of the changes as part of the tvos platform dev.
Have kept this to only ios common changes

## How Has This Been Tested?
Has been in the tvos platform dev code for a few months now
Tested on iOS with no adverse affects.

## Screenshots (if appropriate):

## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [X] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [X] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

sy6sy2 committed:
[iOS/tvOS] Update xcode attributes in Toolchain*.cmake.in

Update xcode attributes for ios/tvos in Toolchain.cmake.in files
Add appletvos platform to Toolchain_binaddons.cmake.in file

fuzzard committed:
[ios/tvos] Implement Codesigning build arguments for darwin_embedded platforms

Add ability to supply codesigning arguments to xcode project build to populate
xcode project after generation.
Refactoring and changes to clean up darwin_embedded archsetup.cmake
Standardise PLATFORM_BUNDLE_IDENTIFIER to allow simple change to xcode project bundle
identifier

fuzzard commented:
jenkins build this please

fuzzard commented:
failures last time round where jenkins crashes

Jenkins build this please

AlwinEsch commented on tools/depends/target/Toolchain_binaddons.cmake.in:
```diff
@@ -37,6 +37,9 @@ elseif(OS STREQUAL darwin_embedded)
   set(CORE_PLATFORM_NAME ios)
+  if(PLATFORM STREQUAL appletvos)
+    set(CORE_PLATFORM_NAME tvos)
+  endif()
```
Maybe:
```cmake
  if(PLATFORM STREQUAL appletvos)
    set(CORE_PLATFORM_NAME tvos)
  else()
    set(CORE_PLATFORM_NAME ios)
  endif()
```

AlwinEsch commented:
Is there something planned to fix the set of `<platform>@PLATFORM@</platform>` on `addon.xml`'s.
On a Mac build comes this currently as `<platform>osx-</platform>` where not usable, it should be `<platform>osx-x86_64</platform>`

Comes e.g. by this addon build way:
```
/Users/Shared/xbmc-depends/x86_64-darwin18.5.0-native/bin/cmake  \
  -DADDONS_TO_BUILD=screensaver.asterwave -DADDON_SRC_PREFIX=../.. \
  -DCMAKE_BUILD_TYPE=Release \
  -DCMAKE_INSTALL_PREFIX="$HOME/Library/Application Support/Kodi/addons" \
  -DPACKAGE_ZIP=1 \
  ../../kodi/cmake/addons`
make
```
The other inside Kodi folder works, but this is the most easiest .

Rechi commented:
The CMake command is wrong, so nothing to fix.

Rechi commented on cmake/scripts/darwin_embedded/ArchSetup.cmake:
```diff
@@ -1,34 +1,37 @@
+  set(ARCH_DEFINES -DTARGET_DARWIN_IOS)
+  set(CMAKE_XCODE_ATTRIBUTE_TARGETED_DEVICE_FAMILY "1,2")
+endif()
+list(APPEND ARCH_DEFINES -DTARGET_POSIX -DTARGET_DARWIN -DTARGET_DARWIN_EMBEDDED)
```
set the common ARCH_DEFINES first, and the add the platform specific ones at the end (move this line above the if block)

Rechi commented on tools/depends/configure.ac:
```diff
@@ -357,14 +357,8 @@ case $host in
+          use_cpu=arm64
           target_platform=appletvos
-          platform_cflags+=" -fembed-bitcode"
-          platform_cxxflags+=" -fembed-bitcode"
```
Why remove adding the `-fembed-bitcode` flags here?

Rechi commented on tools/depends/configure.ac:
```diff
@@ -458,10 +452,12 @@ case $use_platform in
-    fi
-    ;;
+     platform_cflags+=" -fembed-bitcode"
+     platform_cxxflags+=" -fembed-bitcode"
```
Why add adding the `-fembed-bitcode` flags here?

fuzzard replied:
only real reason was trying to split out the platform specific stuff. Started wondering if it was worth adding platform=ios, but never went that route in the end.
Currently ios doesnt use the -fembed-bitcode, but eventually it may i guess.

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -34,25 +34,42 @@ elseif(OS STREQUAL darwin_embedded)
   set(CMAKE_SYSTEM_NAME Darwin)
   set(CORE_SYSTEM_NAME darwin_embedded)
   set(CORE_PLATFORM_NAME ios)
```
move this into the else block of the below if block

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -34,25 +34,42 @@ elseif(OS STREQUAL darwin_embedded)
   # Necessary to build the main Application (but not other dependencies)
   # with Xcode (and a bundle with Makefiles) (https://cmake.org/Bug/view.php?id=15329)
   if(NOT PROJECT_SOURCE_DIR MATCHES "tools/depends")
-    message(STATUS "Toolchain enabled IOS bundle for project ${PROJECT_NAME}")
```
use CORE_PLATFORM_NAME variable in the message

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -34,25 +34,42 @@ elseif(OS STREQUAL darwin_embedded)
+    # complety disable code signing
+    # see: https://gitlab.kitware.com/cmake/cmake/issues/19112
+    set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY "")
+    if(PLATFORM STREQUAL appletvos)
```
compare CORE_PLATFORM_NAME

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -34,25 +34,42 @@ elseif(OS STREQUAL darwin_embedded)
   set(CMAKE_XCODE_ATTRIBUTE_ARCHS ${CPU})
-  set(CMAKE_XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET 9.0)
+  if(CORE_SYSTEM_NAME STREQUAL darwin_embedded)
+    if(PLATFORM STREQUAL appletvos)
```
compare CORE_PLATFORM_NAME

Rechi commented on xbmc/platform/darwin/ios/Info.plist.in:
```diff
@@ -251,7 +251,7 @@
 	</array>
 	<key>CFBundleIdentifier</key>
-	<string>org.xbmc.@APP_NAME_LC@-ios</string>
+	<string>@PLATFORM_BUNDLE_IDENTIFIER@</string>
```
What's the reason to add `-${CORE_PLATFORM_NAME_LC}` at the end of `${APP_PACKAGE}` instead of just using `${APP_PACKAGE}`?

fuzzard replied:
Legacy more than anything to be honest. Thats the layout thats currently used. Defaults back to current setup if you dont provide your own -DBUNDLE_IDENTIFIER_DESC flag. The only other reason would be to differentiate between tvos/ios certs/apps, but come catalina, apple is changing their certs to be generic from what ive read. 
Happy to change to whatever you want really.

kambala-decapitator replied:
`-DPLATFORM_BUNDLE_IDENTIFIER` actually ;)

AlwinEsch commented:
Whats wrong? 
I used this on Windows, Linux and before on Mac and was always OK! Also is this more or less described here: https://github.com/xbmc/xbmc/tree/master/cmake/addons

You mean way by this README.md https://github.com/kodi-pvr/pvr.iptvsimple ?

I went over it, there is really for me a problem. On Linux and Windows, it does not stand out because the cmake variable `CPU` is not needed there.
However, this variable is never set at independent addon build!

Places:
- Here the place where needed:
https://github.com/xbmc/xbmc/blob/master/cmake/scripts/common/PrepareEnv.cmake#L57
- Here the only place where this `CPU` (`WITH_CPU`) variable becomes defined and why only the other way works:
https://github.com/xbmc/xbmc/blob/master/tools/depends/target/Toolchain.cmake.in#L5
This in turn comes from https://github.com/xbmc/xbmc/blob/master/tools/depends/configure.ac

**It can not be that a variable on one side (with CMake) relies on a "configure" on completely different side.**

Here the generated KodiConfig.cmake:
```
set(APP_NAME Kodi)
set(APP_NAME_LC kodi)
set(APP_NAME_UC KODI)
set(APP_PACKAGE org.xbmc.kodi)
set(APP_VERSION_MAJOR 19)
set(APP_VERSION_MINOR 0)
if(NOT KODI_PREFIX)
  set(KODI_PREFIX /Users/alwin/Library/Application Support/Kodi/addons)
endif()
if(NOT KODI_INCLUDE_DIR)
  set(KODI_INCLUDE_DIR /Users/alwin/Development/screensaver.asterwave/build/build/depends/include/kodi)
endif()
if(NOT KODI_LIB_DIR)
  set(KODI_LIB_DIR /Users/alwin/Development/screensaver.asterwave/build/build/depends/lib/kodi)
endif()
if(NOT KODI_DATA_DIR)
  set(KODI_DATA_DIR /Users/alwin/Development/screensaver.asterwave/build/build/depends/share/kodi)
endif()
set(APP_RENDER_SYSTEM )
if(NOT WIN32)
  set(CMAKE_CXX_FLAGS "$ENV{CXXFLAGS} ")
endif()
list(APPEND CMAKE_MODULE_PATH /Users/alwin/Development/screensaver.asterwave/build/build/depends/lib/kodi /Users/alwin/Development/screensaver.asterwave/build/build/depends/share/kodi/cmake)
​
string(REPLACE ";" " " ARCH_DEFINES "-DTARGET_POSIX -DTARGET_DARWIN -DTARGET_DARWIN_OSX")
add_definitions(${ARCH_DEFINES} -DBUILD_KODI_ADDON)
​
set(CORE_SYSTEM_NAME osx)
set(PLATFORM osx)
set(PLATFORM_TAG osx-) <<<-----
set(CMAKE_CXX_STANDARD 11)
set(CMAKE_CXX_STANDARD_REQUIRED ON)
set(CMAKE_CXX_EXTENSIONS OFF)
​
include(AddonHelpers)
```

Also here currently a Hack: https://github.com/xbmc/xbmc/blob/master/cmake/scripts/common/AddonHelpers.cmake#L202

fuzzard commented:
I dont want to derail this too much, but i think what Alwin is getting at is an issue @phunkyfish has spoken to me about on mac. Apparently building just an addon out of tree and then copying into kodi wont run. The cause is probably what Alwin is alluding to here.

I found it odd in iptvsimple that something like https://github.com/kodi-pvr/pvr.iptvsimple/blob/Matrix/build-install-mac.sh was required at all. There is also a bit in the readme that i found wrong https://github.com/kodi-pvr/pvr.iptvsimple/blob/Matrix/README.md#appendix but i have never looked into it yet sorry.

AlwinEsch commented:
Did not want to come across too strict, the current way also simplifies crossplatform construction for iOS, Android etc. I apologize for that.
And I also think it's good to have the way to "/tools/depends/target" and make the automated construction much easier.
I was primarily just to the test and construction on the same system and have therefore made a fix request here: https://github.com/xbmc/xbmc/pull/16547

Rechi commented on tools/depends/configure.ac:
```diff
@@ -357,14 +357,8 @@ case $host in
-            use_cpu=arm64
-          fi
-
+          use_cpu=arm64
```
Why remove the auto cpu selection?

fuzzard replied:
Currently we only have the one cpu for tvos. If apple changes, or someone introduces tvos simulator support (x64), why not just set to what we actually can use

Rechi commented on tools/depends/configure.ac:
```diff
@@ -458,10 +452,12 @@ case $use_platform in
+     if test "$platform_os" != "darwin_embedded"; then
+       AC_MSG_ERROR([$use_platform is only supported on the darwin_embedded platform])
+     fi
+     ;;
```
Why change the indentation level?

fuzzard replied:
Indentation level was just cleaned up to match the other platforms above it.

Rechi commented on tools/depends/configure.ac:
```diff
@@ -687,16 +683,12 @@ AC_SUBST(ffmpeg_options)
 
 if test "$platform_os" == "darwin_embedded"; then
-  if test "$use_platform" = "tvos"; then
-    simulator_sdk_path=[`$use_xcodebuild -version -sdk appletvsimulator$use_sdk  | grep ^Path | awk '{ print $2}'`]
```
Why are you removing this?

fuzzard replied:
We don't actually support building for tvos simulator. If someone does the legwork to introduce it, then we can re add then

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -33,26 +33,41 @@ elseif(OS STREQUAL osx)
+    # complety disable code signing
+    # see: https://gitlab.kitware.com/cmake/cmake/issues/19112
+    set(CMAKE_XCODE_ATTRIBUTE_CODE_SIGN_IDENTITY "")
+    message(STATUS "Toolchain enabled ${CORE_PLATFORM_NAME} bundle for project ${CORE_PLATFORM_NAME}")
```
move the message to it's original place a few line up
`CORE_PLATFORM_NAME` should be `PROJECT_NAME` once

Rechi commented on tools/depends/target/Toolchain.cmake.in:
```diff
@@ -33,26 +33,41 @@ elseif(OS STREQUAL osx)
+  if(CORE_PLATFORM_NAME STREQUAL tvos)
+    set(CMAKE_XCODE_ATTRIBUTE_TVOS_DEPLOYMENT_TARGET 11.0)
+  elseif(CORE_PLATFORM_NAME STREQUAL ios)
+    set(CMAKE_XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET 9.0)
```
Is the if needed?
does this cause issues if `CMAKE_XCODE_ATTRIBUTE_TVOS_DEPLOYMENT_TARGET` is set for an iOS build and `CMAKE_XCODE_ATTRIBUTE_IPHONEOS_DEPLOYMENT_TARGET` for a tvOS build?

fuzzard replied:
More just why add attributes that aren't needed to a project. May handle fine now, but who's to say what will happen in the future. Why not just add what's needed as needed

fuzzard committed:
[ios/tvos] Refactor darwin_embedded platforms in configure.ac

refactor/update configure.ac for ios/tvos platforms
Remove tvos simulator as not implemented at this stage
Update found_sdk grep to better handle finding updated sdk versions.
Update platform_ldflags to point to correct location

fuzzard commented:
Jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
jenkins build this please

fuzzard commented:
Any further comments on this one? I'd like to merge in the next day or two if its ok.

kambala-decapitator commented:
`CFBundleDocumentTypes` key in `Info.plist.in` contains hardcoded strings `org.xbmc.kodi-ios`, I think that should use the resolved bundle ID instead

Rechi commented:
@fuzzard please always set the milestone before or right after merging a PR.
